src/v/kafka/server/handlers/metadata.cc

```diff
@@ -287,6 +287,14 @@ ss::future<response_ptr> metadata_handler::handle(
         }
     }
 
+    // For any edge cases during join or decom where a node might
+    // have an empty list of peers
+    //   related: https://github.com/vectorizedio/redpanda/issues/3030
+    //   related: https://github.com/vectorizedio/redpanda/issues/3289
+    if (reply.data.brokers.empty()) {
+        throw std::runtime_error("No brokers!  This node is in a bad state.");
```

Comment:
Wasn't sure if there was a cleaner way to do an error here other than throwing and letting the rpc layer handle it?  The overall reply didn't look to have an error state.  Each individual topic in the response has an error code, but if we used those then clients might still think the toplevel brokers field was valid.

Reply:
By throwing here I'm pretty sure the server will simply drop the client's connection. If the client would then turn around and retry (which is what I would expect) I'm not sure this would help. OTOH this also seems like the most reasonable thing to do.

Hey @twmb do you have an idea of how redpanda should respond to a metadata request in a case where it doesn't know about any brokers?

A couple questions:

1. Should we always return ourselves as a broker, even if we haven't been able to to join?
2. Perhaps we shouldn't start-up the kafka server layer until we have joined?